python/lsst/daf/butler/registries/sqlPreFlight.py

```diff
@@ -127,29 +180,74 @@ def selectDataUnits(self, collections, expr, neededDatasetTypes, futureDatasetTy
+                for name, col in otherUnit.primaryKeyColumns.items():
+                    _LOG.debug("    joining on column: %s", name)
+                    where.append(dataUnit.table.c[name] == col)
+
+        # joins between skymap and camera units
+        dataUnitJoins = [dataUnitJoin for dataUnitJoin in self._schema.dataUnits.joins.values()
+                         if dataUnitJoin.lhs.issubset(allUnitNames) and
+                         dataUnitJoin.rhs.issubset(allUnitNames)]
+        _LOG.debug("all dataUnitJoins: %s", [join.name for join in dataUnitJoins])
+
+        # only use most specific joins
+        dataUnitJoins = list(_filterSummarizes(dataUnitJoins))
+        _LOG.debug("filtered dataUnitJoins: %s", [join.name for join in dataUnitJoins])
+
+        joinedRegionTables = set()
+        firstRegionIndex = None
+        for dataUnitJoin in dataUnitJoins:
+
+            # TODO: do not know yet how to handle MultiCameraExposureJoin, skip it for now
+            if dataUnitJoin.lhs == dataUnitJoin.rhs:
+                continue
+
+            for connection in (dataUnitJoin.lhs, dataUnitJoin.rhs):
+                regionDataUnit = self._schema.dataUnits.getRegionDataUnit(*connection)
+                if len(connection) > 1:
+                    # if one of the joins is with Visit/Sensor then also bring
+                    # VisitSensorRegion table in and join it with the units
+                    if regionDataUnit.name in joinedRegionTables:
+                        _LOG.debug("region table already joined with units: %s", regionDataUnit.name)
+                    else:
+                        _LOG.debug("joining region table with units: %s", regionDataUnit.name)
+                        joinedRegionTables.add(regionDataUnit.name)
+
+                        for dataUnitName in connection:
+                            dataUnit = self._schema.dataUnits[dataUnitName]
+                            _LOG.debug("  joining region table with %s", dataUnitName)
+                            for name, col in dataUnit.primaryKeyColumns.items():
+                                _LOG.debug("    joining on column: %s", name)
+                                where.append(regionDataUnit.table.c[name] == col)
+
+                # now join region table with join table using PKs of all units
+                _LOG.debug("join %s with %s", dataUnitJoin.name, connection)
+                for colName in self._schema.dataUnits.getPrimaryKeyNames(connection):
+                    _LOG.debug("  joining on column: %s", colName)
+                    where.append(dataUnitJoin.table.c[colName] == regionDataUnit.table.c[colName])
+
+                # We also have to include regions from each side of the join into
+                # resultset so that we can filter-out non-overlapping regions.
+                firstRegionIndex = len(header)
+                selectColumns.append(regionDataUnit.regionColumn)
```

Comment:
It would be nice (not necessarily for this ticket) to make the extra filtering based on region configurable via an option in schema.yaml, for databases that can do that inside the query (and define the spatial join tables accordingly) - I think this is actually doable for SQLite (see [DM-14830](https://jira.lsstcorp.org/browse/DM-14830)), though I don't know if it will be efficient.

Reply:
We'll have to experiment with that. For SQLite I do not think we'll see big difference as everything is still done on client side. It may make difference for Oracle but I have no idea how to implement server-side functions that know about `sphgeom` regions.